physicsnemo/sym/eq/ls/grads.py

```diff
@@ -23,148 +23,189 @@
-        p2 = coords[connectivity_tensor[:, :, 1]]
-        dx = p1[:, :, 0] - p2[:, :, 0]
-        dy = p1[:, :, 1] - p2[:, :, 1]
+    def forward(self, coords, connectivity_tensor, y, du=None, dv=None) -> List[Tensor]:
+        """
+        Compute first derivatives using least squares method with fully vectorized computation.
+
+        Parameters
+        ----------
+        coords : torch.Tensor
+            Node coordinates of shape [N, dim] (ignored if direct_input=True)
+        connectivity_tensor : tuple[torch.Tensor, torch.Tensor] or tuple[torch.Tensor, torch.Tensor, torch.Tensor]
+            Either (offsets, indices) for sparse format or (offsets, indices, neighbor_matrix) for batched format (ignored if direct_input=True)
+        y : torch.Tensor
+            Function values at nodes of shape [N, 1] (ignored if direct_input=True)
+        du : torch.Tensor, optional
+            Direct input for function differences, shape [N, max_neighbors, 1] or [batch_size, N, max_neighbors, 1] (required if direct_input=True)
+        dv : torch.Tensor, optional
+            Direct input for coordinate differences, shape [N, max_neighbors, dim] or [batch_size, N, max_neighbors, dim] (required if direct_input=True)
+
+        Returns
+        -------
+        List[torch.Tensor]
+            List of gradient components [dudx, dudy, dudz] for each node
+        """
+        if self.direct_input:
+            if du is None or dv is None:
+                raise ValueError("du and dv must be provided when direct_input=True")
+            return self._forward_direct(du, dv)
+
+        # Handle different connectivity formats
+        if len(connectivity_tensor) == 2:
+            offsets, indices = connectivity_tensor
+            return self._forward_sparse(coords, offsets, indices, y)
+        elif len(connectivity_tensor) == 3:
+            _, _, neighbor_matrix = connectivity_tensor
+            return self._forward_batched(coords, neighbor_matrix, y)
+        else:
+            raise ValueError("connectivity_tensor must be tuple of length 2 or 3")
+
+    def _forward_direct(self, du, dv) -> List[Tensor]:
+        """
+        Compute derivatives directly from provided du and dv tensors.
+
+        Parameters
+        ----------
+        du : torch.Tensor
+            Function differences, shape [N, max_neighbors, 1] or [batch_size, N, max_neighbors, 1]
+        dv : torch.Tensor
+            Coordinate differences, shape [N, max_neighbors, dim] or [batch_size, N, max_neighbors, dim]
+
+        Returns
+        -------
+        List[torch.Tensor]
+            List of gradient components [dudx, dudy, dudz] for each node
+        """
+
+        grad_u = self.compute_ls_grads(dv, du)  # [batch_size, N, dim, 1]
+        grad_u = grad_u.squeeze(-1)  # [batch_size, N, dim]
+
+        # Split into individual components
+        result = []
+        for i in range(self.dim):
+            result.append(grad_u[:, i : i + 1])
```

Comment:
Not a critical change, but for best-practices, this should be rewritten with a list comprehension. (For example, Ruff would flag this as a performance error)